chips/cc26xx/src/trng.rs

```diff
@@ -0,0 +1,158 @@
+    cfg0: ReadWrite<u32, Config::Register>,
+
+    alarm_ctl: ReadWrite<u32, AlarmControl::Register>,
+
+    _fro_en: ReadOnly<u32>,
+    _fro_detune: ReadOnly<u32>,
+
+    _alarm_mask: ReadOnly<u32>,
+    _alarm_stop: ReadOnly<u32>,
+
+    _lfsr0: ReadOnly<u32>,
+    _lfsr1: ReadOnly<u32>,
+    _lfsr2: ReadOnly<u32>,
+
+    _r0: [u8; 0x1FB4],
+
+    sw_reset: ReadWrite<u32, SoftwareReset::Register>,
+}
+
+register_bitfields![
+    u32,
+    IrqStatus [
+        READY OFFSET(0) NUMBITS(1) []
+    ],
+    IrqFlagClear [
+        READY OFFSET(0) NUMBITS(1) []
+    ],
+    Control [
+        STARTUP_CYCLES  OFFSET(16) NUMBITS(16)  [],
+        TRNG_EN         OFFSET(10) NUMBITS(1)   []
+    ],
+    Config [
+        MAX_REFILL_CYCLES   OFFSET(16) NUMBITS(16) [],
+        SMPL_DIV            OFFSET(8)  NUMBITS(4)  [],
+        MIN_REFILL_CYCLES   OFFSET(0)  NUMBITS(8)  []
+    ],
+    AlarmControl [
+        // Alarm threshold for repeating pattern detectors
+        ALARM_THR   OFFSET(0)   NUMBITS(8) []
+    ],
+    SoftwareReset [
+        RESET OFFSET(0) NUMBITS(1) []
+    ]
+];
+
+const BASE_ADDRESS: *mut RngRegisters = 0x4002_8000 as *mut RngRegisters;
+
+pub static mut TRNG: Trng = Trng::new();
+
+pub struct Trng {
+    regs: *mut RngRegisters,
+    client: Cell<Option<&'static rng::Client>>,
+}
+
+impl Trng {
+    const fn new() -> Trng {
+        Trng {
+            regs: BASE_ADDRESS,
+            client: Cell::new(None),
+        }
+    }
+
+    pub fn enable(&self) {
+        // Ensure that the power domain TRNG resides in is enabled
+        if !prcm::Power::is_enabled(prcm::PowerDomain::Peripherals) {
+            prcm::Power::enable_domain(prcm::PowerDomain::Peripherals);
+
+            while !prcm::Power::is_enabled(prcm::PowerDomain::Peripherals) {}
+        }
+
+        // Setup the clock
+        prcm::Clock::enable_trng();
+
+        let regs = unsafe { &*self.regs };
+
+        regs.ctl.set(0);
+
+        // Issue a SW reset
+        regs.sw_reset.write(SoftwareReset::RESET::SET);
+        while !regs.sw_reset.is_set(SoftwareReset::RESET) {}
+
+        // Set the startup samples
+        regs.ctl.modify(Control::STARTUP_CYCLES.val(1));
+
+        // Configure the minimum and maximum number of samples per generated number
+        // and the number of clock cycles per sample.
+        //  NOTE: tune these if the generation is not satisfactory
+        let max_samples_per_cycle = 0x100;
+        let min_samples_per_cycle = 0;
+        let cycles_per_sample = 0;
+        regs.cfg0.write(
+            Config::MAX_REFILL_CYCLES.val(max_samples_per_cycle >> 8)
+                + Config::SMPL_DIV.val(cycles_per_sample)
+                + Config::MIN_REFILL_CYCLES.val(min_samples_per_cycle >> 6),
+        );
+
+        // Reset the alarm control
+        regs.alarm_ctl.write(AlarmControl::ALARM_THR.val(0xFF));
+
+        // Enable the TRNG
+        regs.ctl.modify(Control::TRNG_EN::SET);
+    }
+
+    pub fn read_number(&self) -> u64 {
```

Comment:
add a comment or change function to `read_number_blocking`

Reply:
Also why u64 here, the capsule expects randomness in chunks of 4 bytes.

Reply:
A function rename seems like a good idea!

Seemed reasonable at the time to use a u64 since the hardware is capable of generating two 32-bit integers, and it's possible to use both of them this way without an extra function call (eg. `read_number0`, `read_number1`)